### PR TITLE
Add more events

### PR DESCRIPTION
**Pull request by KingRainbow44**
## Description

This PR adds more events for plugins to use, along with some basic implementations.\
Some events (like move event) could be better polished but I would save that for later.

- [ ] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by KingRainbow44** [comment]
will remain draft until @Melledy finishes his merge + I fix teleport command

**Review by ghost — APPROVED** [review]

**Review by Birdulon — COMMENTED** [review]
I don't really like how every Teleport action now needs to have all these additional calls.
Maybe it should just be a helper function somewhere that runs the event check, performs the teleport, and returns a success boolean for any reporting purposes.

**Comment by ghost** [comment]
> I don't really like how every Teleport action now needs to have all these additional calls. Maybe it should just be a helper function somewhere that runs the event check, performs the teleport, and returns a success boolean for any reporting purposes.

I think a better way is to put the teleport event inside `World::transferPlayerToScene()` so there is no need to add the teleport event call before the actual teleport function is called.

**Comment by Birdulon** [comment]
> I think a better way is to put the teleport event inside `World::transferPlayerToScene()`

Maybe throw a custom exception type from `World::transferPlayerToScene()` to distinguish it from failing because of invalid scene?

**Comment by KingRainbow44** [comment]
Will implement the event inside of `World#transferPlayerToScene` with a custom exception for errors.

**Comment by KingRainbow44** [comment]
> with a custom exception for errors

_totally didn't forget to add this_

File: src/main/java/emu/grasscutter/server/packet/recv/HandlerTryEnterHomeReq.java [hunk 1]
```diff
@@ -38,13 +35,9 @@ public void handle(GameSession session, byte[] header, byte[] payload) throws Ex
...
-                realmId,
-                pos
-        );
-
-
-        session.send(new PacketTryEnterHomeRsp(req.getTargetUid()));
+        boolean result = session.getPlayer().getWorld().transferPlayerToScene(
+            session.getPlayer(), realmId,
+            TeleportType.WAYPOINT, pos
+        ); if (result) session.send(new PacketTryEnterHomeRsp(req.getTargetUid()));
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/recv/HandlerTryEnterHomeReq.java]:
```suggestion
        );
        if (result) session.send(new PacketTryEnterHomeRsp(req.getTargetUid()));
```

**Review by ghost — APPROVED** [review]

File: src/main/java/emu/grasscutter/game/managers/mapmark/MapMarksManager.java [hunk 2]
```diff
@@ -30,54 +30,51 @@ public void handleMapMarkReq(MarkMapReq req) {
...
+        if (this.getMapMarks().size() < mapMarkMaxCount) {
+            this.getMapMarks().put(this.getMapMarkKey(mapMark.getPosition()), mapMark);
         }
     }
 
     private void teleport(Player player, MapMark mapMark) {
-        float y;
-        try {
+        float y; try {
             y = (float)Integer.parseInt(mapMark.getName());
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/mapmark/MapMarksManager.java]:
```suggestion
        float y;
        try {
            y = Float.parseFloat(mapMark.getName());
```

File: src/main/java/emu/grasscutter/server/packet/recv/HandlerSceneTransToPointReq.java [hunk 3]
```diff
@@ -1,36 +1,36 @@
...
+		String code = req.getSceneId() + "_" + req.getPointId();
+		ScenePointEntry scenePointEntry = GameData.getScenePointEntries().get(code);
+
+		if (scenePointEntry != null) {
+			float x = scenePointEntry.getPointData().getTranPos().getX();
+			float y = scenePointEntry.getPointData().getTranPos().getY();
+			float z = scenePointEntry.getPointData().getTranPos().getZ();
+
+            boolean result = session.getPlayer().getWorld().transferPlayerToScene(session.getPlayer(), req.getSceneId(), TeleportType.WAYPOINT, new Position(x, y, z));
+            if(result) session.send(new PacketSceneTransToPointRsp(session.getPlayer(), req.getPointId(), req.getSceneId()));
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/recv/HandlerSceneTransToPointReq.java]:
```suggestion
            if (session.getPlayer().getWorld().transferPlayerToScene(session.getPlayer(), req.getSceneId(), TeleportType.WAYPOINT, new Position(x, y, z))) {
                session.send(new PacketSceneTransToPointRsp(session.getPlayer(), req.getPointId(), req.getSceneId()));
```

**Review by Birdulon — APPROVED** [review]
I'm not really huge on explicit `this`, but otherwise no complaints apart from the merge conflicts.

**Pull request merged**